Demonstrate augmented Lagrange preconditioning for Taylor Hood elements

lindsayad opened the pull request:
Refs #27126

lindsayad committed:
Demonstrate augmented Lagrange preconditioning for Taylor Hood elements

Refs #27126

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2126403/) on eeb8478 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27127/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27127/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f5ce2c7ab1f32ca0867aa104c576fca287b26ed0`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2126458/) on d595970 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27127/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2126502/) on d595970 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f5ce2c</th>
      <th colspan=4>#27127 d59597</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.31%</td>
      <td>85.31%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>101273</td>
      <td>101272</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17443</td>
      <td>17444</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27127/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27127/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>f5ce2c</th>
      <th colspan=4>#27127 d59597</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.80%</td>
      <td>84.80%</td>
      <td>+0.01%</td>
      <td>95.45%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14584</td>
      <td>14600</td>
      <td>+16</td>
      <td>21</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2615</td>
      <td>2616</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/27127/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27127/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27127/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on modules/navier_stokes/doc/content/bib/navier_stokes.bib:
```diff
@@ -422,3 +422,34 @@ @article{zanetti2020scalable
+}
+
+@article{benzi2023solving,
+  title={Solving linear systems of the form by preconditioned iterative methods},
```
title didnt paste well

GiudGiud commented on modules/navier_stokes/include/base/NavierStokesMethods.h:
```diff
@@ -148,4 +148,22 @@ void getElementFaceArgs(const std::vector<BoundaryName> & wall_boundary_name,
+  if (coord_sys == Moose::COORD_RZ)
+    // u_r / r
+    div += value(rz_radial_coord) / point(rz_radial_coord);
+  return div;
```
mooseAssert not spherical? or mooseAssert XYZ or RZ

GiudGiud commented on modules/navier_stokes/include/kernels/INSADMomentumGradDiv.h:
```diff
@@ -0,0 +1,34 @@
+  /// The grad-div stabilization coefficient
+  const Real _gamma;
+
+  const Moose::CoordinateSystemType & _coord_sys;
```
```suggestion
  /// Coordinate system of the mesh
  const Moose::CoordinateSystemType & _coord_sys;
```

GiudGiud commented on modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md:
```diff
@@ -0,0 +1,55 @@
+
+where $\nu$ is the kinematic viscosity and $M_p$ is the pressure mass matrix
+([!citep](farrell2019augmented)). In [!citep](benzi2006augmented) the authors
+found that $\gamma$ on the order of the velocity magnitude provides $\nu$- and
+$h$-independent (the latter denoting element size) convergence for a wide range
```
so gamma is ok as a constant? or do they make it a field

lindsayad replied:
it should be a constant

GiudGiud reviewed (commented):
looks pretty cool. I ll try to read one of the 3 to get up to speed

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/lid_driven/steady_vector_fsp_al.i:
```diff
@@ -0,0 +1,165 @@
+    type = INSADMomentumGradDiv
+    variable = vel
+    gamma = ${gamma}
+  []
```
should this contribute to the residual too?
I guess I ll know once I read the paper but I was expecting it from the doco to only contribute to the mass matrix

GiudGiud replied:
yup I see it equation 2.6 of Farrell

grmnptr commented on modules/navier_stokes/doc/content/source/kernels/INSADMomentumGradDiv.md:
```diff
@@ -0,0 +1,55 @@
+In this case we only wish to display the effectiveness of the pressure mass
+matrix as a preconditioner for the Schur complement, so we use LU solves for
+inverting the velocity block and $M_p$. Indeed for $\nu=10^{-3}$ (equivalent to
+$\textrm{Re}=1000$), the solution of the Schur complement system takes no more
```
```suggestion
$\textrm{Re}=1000$), the solution of the Schur complement system to a relative tolerance of ($10^{-4}$) takes no more
```

grmnptr commented on modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests:
```diff
@@ -96,13 +96,23 @@ finite elements for velocity and pressure.'
+      input = 'steady_vector_fsp_al.i'
+      exodiff = 'steady_vector_fsp_al_exo.e'
+      petsc_version = '>=3.20.0'
+      detail = 'a high Reynolds number (100) Navier-Stokes problem with a grad-div stabilization.'
```
```suggestion
      detail = 'a high Reynolds number (1000) Navier-Stokes problem with a grad-div stabilization.'
```

grmnptr reviewed (changes requested):
Nice new feature!

lindsayad committed:
Apply suggestions from code review

- specify tolerance to which we are solving the Schur complement
- fix the Reynolds number in the SQA requirement
- fix title in bib
- add doxygen for coordinate system
- assert Cartesian or RZ coordinate system

Co-authored-by: Peter German <31662443+grmnptr@users.noreply.github.com>
Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
failures are unrelated

lindsayad commented:
thanks guys for the quick reviews!
